Make UserMediaPermissionRequestManagerProxy refcounted since it subclasses CanMakeWeakPtr

cdumez opened the pull request:
#### c46a07050a1451a5c58ac375f805d8b3582510c8
<pre>
Make UserMediaPermissionRequestManagerProxy refcounted since it subclasses CanMakeWeakPtr
<a href="https://bugs.webkit.org/show_bug.cgi?id=280158">https://bugs.webkit.org/show_bug.cgi?id=280158</a>

Reviewed by Youenn Fablet.

UserMediaPermissionRequestManagerProxy subclasses CanMakeWeakPtr so it should be
ref-counted too, for safety.

* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::ref const):
(WebKit::UserMediaPermissionRequestManagerProxy::deref const):
(WebKit::UserMediaPermissionRequestManagerProxy::captureDevicesChanged):
(WebKit::UserMediaPermissionRequestManagerProxy::grantRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::finishGrantingRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::startProcessingUserMediaPermissionRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::processUserMediaPermissionRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::processUserMediaPermissionValidRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::getUserMediaPermissionInfo):
(WebKit::UserMediaPermissionRequestManagerProxy::computeFilteredDeviceList):
(WebKit::UserMediaPermissionRequestManagerProxy::enumerateMediaDevicesForFrame):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.h:
* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.cpp:
(WebKit::UserMediaPermissionRequestProxy::UserMediaPermissionRequestProxy):
(WebKit::UserMediaPermissionRequestProxy::manager const):
(WebKit::UserMediaPermissionRequestProxy::allow):
(WebKit::UserMediaPermissionRequestProxy::deny):
(WebKit::UserMediaPermissionRequestProxy::isPending const):
(WebKit::UserMediaPermissionRequestProxy::promptForGetDisplayMedia):
(WebKit::UserMediaPermissionRequestProxy::promptForGetUserMedia):
* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.h:
(WebKit::UserMediaPermissionRequestProxy::isPending const): Deleted.
(WebKit::UserMediaPermissionRequestProxy::manager const): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::userMediaPermissionRequestManager):

Canonical link: <a href="https://commits.webkit.org/284117@main">https://commits.webkit.org/284117@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43a0dcfa5b231d344faa09919d38ea9898ad3911

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68473 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47865 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21132 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72542 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19618 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55661 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19434 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13066 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71540 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43757 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59125 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/35124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40425 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17975 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62384 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16893 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74236 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12444 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16162 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62113 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12483 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59203 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62137 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10066 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3680 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10428 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43666 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44740 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45934 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44482 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
```diff
@@ -122,6 +122,16 @@ UserMediaPermissionRequestManagerProxy::~UserMediaPermissionRequestManagerProxy(
 
+void UserMediaPermissionRequestManagerProxy::ref() const
+{
+    m_page->ref();
```
Refing the page does not guarantee that its UserMediaPermissionRequestManagerProxy member will not be set to null (this happens when the page is reset).
It is probably safer to make it RefCounted.

cdumez replied:
Oh, I missed that. Will fix.

youennf commented on Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
```diff
@@ -307,8 +323,19 @@ void UserMediaPermissionRequestManagerProxy::finishGrantingRequest(UserMediaPerm
+
+    page->willStartCapture(request, [this, weakThis = WeakPtr { *this }, request = Ref { request }]() mutable {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis)
```
In theory, we should probably deny the request here as well since it owns a CompletionHandler.
Or maybe have UserMediaPermissionRequestProxy call its completion handler (if any) in its destructor as a catch them all handling.

youennf commented on Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
```diff
@@ -318,20 +345,20 @@ void UserMediaPermissionRequestManagerProxy::finishGrantingRequest(UserMediaPerm
-            if (!--m_hasPendingCapture)
-                UserMediaProcessManager::singleton().revokeSandboxExtensionsIfNeeded(this->page().protectedLegacyMainFrameProcess());
+            if (!--m_hasPendingCapture && this->page())
+                UserMediaProcessManager::singleton().revokeSandboxExtensionsIfNeeded(this->page()->protectedLegacyMainFrameProcess());
```
We probably should ref page() here.

youennf commented on Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.mm:
```diff
@@ -86,7 +91,7 @@
 #if ENABLE(MEDIA_STREAM)
     auto overridePreference = DisplayCaptureSessionManager::singleton().overrideCanRequestDisplayCapturePermissionForTesting();
-    if (!manager() || (!overridePreference && manager()->page().preferences().requireUAGetDisplayMediaPrompt()))
+    if (!manager() || !manager()->page() || (!overridePreference && manager()->page()->preferences().requireUAGetDisplayMediaPrompt()))
```
I would add a protectedPage variable here.

youennf commented on Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.mm:
```diff
@@ -51,7 +51,8 @@
 #if ENABLE(MEDIA_STREAM)
     if (m_hasPendingGetDispayMediaPrompt) {
-        DisplayCaptureSessionManager::singleton().cancelGetDisplayMediaPrompt(Ref { manager()->page() });
+        if (RefPtr page = manager()->page())
```
In theory, we should add a protectedManager() here.

youennf commented on Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.mm:
```diff
@@ -64,8 +65,12 @@
     if (!manager())
         return;
 
+    RefPtr page = manager()->page();
```
I would tend to do a RefPtr manager variable here.

cdumez committed:
Make UserMediaPermissionRequestManagerProxy refcounted since it subclasses CanMakeWeakPtr
https://bugs.webkit.org/show_bug.cgi?id=280158

Reviewed by Youenn Fablet.

UserMediaPermissionRequestManagerProxy subclasses CanMakeWeakPtr so it should be
ref-counted too, for safety.

* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::ref const):
(WebKit::UserMediaPermissionRequestManagerProxy::deref const):
(WebKit::UserMediaPermissionRequestManagerProxy::captureDevicesChanged):
(WebKit::UserMediaPermissionRequestManagerProxy::grantRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::finishGrantingRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::startProcessingUserMediaPermissionRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::processUserMediaPermissionRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::processUserMediaPermissionValidRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::getUserMediaPermissionInfo):
(WebKit::UserMediaPermissionRequestManagerProxy::computeFilteredDeviceList):
(WebKit::UserMediaPermissionRequestManagerProxy::enumerateMediaDevicesForFrame):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.h:
* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.cpp:
(WebKit::UserMediaPermissionRequestProxy::UserMediaPermissionRequestProxy):
(WebKit::UserMediaPermissionRequestProxy::manager const):
(WebKit::UserMediaPermissionRequestProxy::allow):
(WebKit::UserMediaPermissionRequestProxy::deny):
(WebKit::UserMediaPermissionRequestProxy::isPending const):
(WebKit::UserMediaPermissionRequestProxy::promptForGetDisplayMedia):
(WebKit::UserMediaPermissionRequestProxy::promptForGetUserMedia):
* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.h:
(WebKit::UserMediaPermissionRequestProxy::isPending const): Deleted.
(WebKit::UserMediaPermissionRequestProxy::manager const): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::userMediaPermissionRequestManager):

Canonical link: https://commits.webkit.org/284117@main
